Pass current user to sharing calls in shared storage/cache

PVince81 opened the pull request:
In some situations like when running background jobs, there is no logged
in user. However the filesystem has been setup for a given user to
perform operations on it. That user must be used when calling the Share
API instead of letting it rely on the logged in user.

Fixes https://github.com/owncloud/core/issues/15852
- [x] TODO: add unit tests
- [x] requires https://github.com/owncloud/core/issues/15867 to be fixed to be able to check if expiry is done properly after move

@schiesbn @icewind1991

Commit:
Pass current user to sharing calls in shared storage/cache

In some situations like when running background jobs, there is no logged
in user. However the filesystem has been setup for a given user to
perform operations on it. That user must be used when calling the Share
API instead of letting it rely on the logged in user.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ed02277b-615e-4f64-8db5-939119750248) was created.

icewind1991 commented on lib/private/share/share.php:
```diff
@@ -281,12 +281,31 @@ public static function getUsersSharingFile($path, $ownerUser, $includeOwner = fa
+			$itemType,
+			null,
+			self::$shareTypeUserAndGroups,
+			\OC_User::getUser(),
```
`$recipientUser` ?

PVince81 replied:
nice catch! Thanks.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/ed02277b-615e-4f64-8db5-939119750248) completed: **3 new issues, 7 updated code elements**

nickvergessen commented on apps/files_versions/lib/storage.php:
```diff
@@ -67,10 +67,26 @@ class Storage {
+	 *
+	 * @param string $filename path to the file name to resolve, relative to
+	 * the current user's home
+	 * @param string $currentUser current user, defaults to the currently logged
```
I don't think the variable name `currentUser` makes a lot of sense, when it might not be the current user, just `user` instead?

nickvergessen commented on apps/files_versions/lib/storage.php:
```diff
@@ -67,10 +67,26 @@ class Storage {
+	 */
+	public static function getUidAndFilename($filename, $currentUser = null) {
+		if ($currentUser === null) {
+			$currentUser = \OC::$server->getUserSession()->getUser()->getUID();
```
`getUser()` might return null, please catch this case

PVince81 replied:
I know... just didn't want to... because if $user is null it will fail anyway. Maybe need to add a warning then and bail out.

nickvergessen commented on apps/files_versions/lib/storage.php:
```diff
@@ -500,15 +516,18 @@ private static function scheduleExpire($fileName, $versionsSize = null, $neededS
-	 * @param int|null $versionsSize
-	 * @param int $offset
-	 * @return bool|int|null
+	 * @param $currentUser current user
```
variable name here again, it can be any user, doesn't have to be the current one

nickvergessen commented on lib/private/share/share.php:
```diff
@@ -281,12 +281,31 @@ public static function getUsersSharingFile($path, $ownerUser, $includeOwner = fa
 	 * @param mixed $parameters (optional)
 	 * @param int $limit Number of items to return (optional) Returns all by default
 	 * @param boolean $includeCollections (optional)
+	 * @param string $recipientUser id of the recipient user (optional), defaults
```
incomplete `... defaults [to current user]`?

nickvergessen commented on lib/public/share.php:
```diff
@@ -95,13 +96,26 @@ public static function getUsersSharingFile($path, $ownerUser, $includeOwner = fa
 	 * @param mixed $parameters (optional)
 	 * @param int $limit Number of items to return (optional) Returns all by default
 	 * @param bool $includeCollections (optional)
+	 * @param string $recipientUser id of the recipient user (optional), defaults
```
same here, you copied the doc line incomplete from the other method

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/07b44bed-6cb4-462e-a3a3-4fdffc2abe04) was created.

PVince81 commented:
I fixed the comments and also added a unit test.
That unit test will not be able to pass until we fix https://github.com/owncloud/core/issues/15867 to properly move the versions. I'll have a look.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/07b44bed-6cb4-462e-a3a3-4fdffc2abe04) completed: **4 new issues, 8 updated code elements**

nickvergessen commented:
Testergebnis (2 fehlgeschlagene Tests / +2)

    Test_Files_Versioning.testRenameInSharedFolder
    Test_Files_Versioning.testMoveIntoSharedFolderAsRecipient

PVince81 commented:
Yes, expected: https://github.com/owncloud/core/pull/15868#issuecomment-96584625

PVince81 commented:
The original issue doesn't happen any more, see https://github.com/owncloud/core/issues/15852#issuecomment-102432379

However I'd like to keep this PR for 8.2 as it removes the reliance on the currently logged in user and makes it possible to pass a target user to some share APIs.

PVince81 commented:
Will continue working on this for 8.2.
The following unit tests will fail:
```
1) Test_Files_Versioning::testMoveIntoSharedFolderAsRecipient
Failed asserting that false is true.

/srv/www/htdocs/owncloud/apps/files_versions/tests/versions.php:485

2) Test_Files_Versioning::testStoreVersionAsAnonymous
Failed asserting that 0 is greater than 0.

/srv/www/htdocs/owncloud/apps/files_versions/tests/versions.php:781
/srv/www/htdocs/owncloud/apps/files_versions/tests/versions.php:761
```

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/8b28b5cb-809a-4ef1-a1f0-8d0139710e18) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/bd0148dd-51e0-41ad-871b-df6f27ad5d8e) was created.

PVince81 commented:
Rebased.

This is the remains of an older obsoleted PR. Basically the changes here aren't strictly required anymore to fix expiration, I managed to fix it differently.

However, this PR brings the advantage of being able to pass the current user to various sharing functions, which could be useful at some point and is cleaner.

@schiesbn @icewind1991 what do you think ? Is it worth it ?

PVince81 commented:
Sharing will be revamped for 2.0 => close
